builtin/providers/azurerm/resource_arm_virtual_machine_extension.go

```diff
@@ -0,0 +1,232 @@
+	client := meta.(*ArmClient).vmExtensionClient
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	vmName := d.Get("virtual_machine_name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	publisher := d.Get("publisher").(string)
+	extensionType := d.Get("type").(string)
+	typeHandlerVersion := d.Get("type_handler_version").(string)
+	autoUpgradeMinor := d.Get("auto_upgrade_minor_version").(bool)
+	tags := d.Get("tags").(map[string]interface{})
+
+	settings, err := expandArmVirtualMachineExtensionSettings(d.Get("settings").(string))
+	if err != nil {
+		return fmt.Errorf("unable to parse settings: %s", err)
+	}
+
+	extension := compute.VirtualMachineExtension{
+		Location: &location,
+		Properties: &compute.VirtualMachineExtensionProperties{
+			Publisher:               &publisher,
+			Type:                    &extensionType,
+			TypeHandlerVersion:      &typeHandlerVersion,
+			AutoUpgradeMinorVersion: &autoUpgradeMinor,
+			Settings:                &settings,
+		},
+		Tags: expandTags(tags),
+	}
+
+	if protectedSettingsString := d.Get("protected_settings").(string); protectedSettingsString != "" {
+		protectedSettings, err := expandArmVirtualMachineExtensionSettings(protectedSettingsString)
+		if err != nil {
+			return fmt.Errorf("unable to parse protected_settings: %s", err)
+		}
+		extension.Properties.ProtectedSettings = &protectedSettings
+	}
+
+	_, err = client.CreateOrUpdate(resGroup, vmName, name, extension, make(chan struct{}))
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(resGroup, vmName, name, "")
+	if err != nil {
+		return err
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read  Virtual Machine Extension %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmVirtualMachineExtensionsRead(d, meta)
+}
+
+func resourceArmVirtualMachineExtensionsRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).vmExtensionClient
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	vmName := id.Path["virtualMachines"]
+	name := id.Path["extensions"]
+
+	resp, err := client.Get(resGroup, vmName, name, "")
+	if err != nil {
+		return fmt.Errorf("Error making Read request on Virtual Machine Extension %s: %s", name, err)
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("location", azureRMNormalizeLocation(*resp.Location))
+	d.Set("virtual_machine_name", vmName)
+	d.Set("resource_group_name", resGroup)
+	d.Set("publisher", resp.Properties.Publisher)
+	d.Set("type", resp.Properties.Type)
+	d.Set("type_handler_version", resp.Properties.TypeHandlerVersion)
+	d.Set("auto_upgrade_minor_version", resp.Properties.AutoUpgradeMinorVersion)
+
+	settings, err := flattenArmVirtualMachineExtensionSettings(*resp.Properties.Settings)
```

Comment:
Seems that `settings` is optional as far as the API is concerned, I thought it was required, I'll guard against a panic in the read since it's the only settings map read back.